metron-analytics/metron-profiler/src/main/java/org/apache/metron/profiler/bolt/ProfileBuilderBolt.java

```diff
@@ -310,17 +313,37 @@ public void execute(TupleWindow window) {
   }
 
   /**
-   * Flush all expired profiles when a 'tick' is received.
+   * Flush all active profiles.
+   */
+  protected void flushActive() {
+    activeFlushSignal.reset();
+
+    // flush the active profiles
+    List<ProfileMeasurement> measurements;
+    synchronized(messageDistributor) {
+      measurements = messageDistributor.flush();
+      emitMeasurements(measurements);
+    }
+
+    LOG.debug("Flushed active profiles and found {} measurement(s).", measurements.size());
+
+  }
+
+  /**
+   * Flushes all expired profiles.
    *
-   * If a profile has not received a message for an extended period of time then it is
+   * <p>If a profile has not received a message for an extended period of time then it is
    * marked as expired.  Periodically we need to flush these expired profiles to ensure
    * that their state is not lost.
    */
-  private void handleTick() {
+  protected void flushExpired() {
 
     // flush the expired profiles
-    List<ProfileMeasurement> measurements = messageDistributor.flushExpired();
-    emitMeasurements(measurements);
+    List<ProfileMeasurement> measurements;
+    synchronized (messageDistributor) {
```

Comment:
Access to the `messageDistributor` has to be synchronized now.  It is not thread safe and it could be called from either the timer thread or when tuples are received.